lab2/lab2.md

```diff
@@ -1,30 +1,755 @@
+You can think of them as "unconditional branches."
+
+## `jal` instruction details
+
+The following table shows how the `jal` instruction is laid out.
+
+| 31-12                    | 11-7 | 6-0     | Name   |
+|--------------------------|------|---------|--------|
+| imm[20, 10:1, 11, 19:12] | rd   | 1101111 | jal    |
+
+`jal` stands for "jump and link."
+The instruction has the following effect.
+
+```
+pc = pc + imm
+R[rd] = pc + 4
+```
+
+## Testing `jal`
+
+You can run the tests for this part with the following command:
+
+```
+sbt:dinocpu> testOnly dinocpu.SingleCycleJALTesterLab2
+```
+
+# Part IX: `jalr`
+
+`jalr` is very similar to `jal`, with one difference.
+However, unlike `jal`, `jalr` has the format of an I-type instruction.
+
+## `jalr` instruction details
+
+The following table shows how the `jalr` instruction is laid out.
+
+| 31-20     | 19-15 | 14-12 | 11-7 | 6-0     | Name   |
+|-----------|-------|-------|------|---------|--------|
+| imm[11:0] | rs1   | 000   | rd   | 1100111 | jalr   |
+
+`jalr` stands for "jump and link register.""
+The instruction has the following effect.
+(Careful, there's one major difference between this and `jal`!)
+
+```
+pc = R[rs1] + imm
+R[rd] = pc + 4
+```
+
+## Testing `jalr`
+
+You can run the tests for this part with the following command:
+
+```
+sbt:dinocpu> testOnly dinocpu.SingleCycleJALRTesterLab2
+```
+
+# Part X: Full applications
+
+At this point, you should have a fully implemented RISC-V CPU!
+In this final part of the assignment, you will run some full RISC-V applications.
+
+We have provided four applications for you.
+
+* `fibonacci`, which computes the nth Fibonacci number. The initial value of `t1` contains the Fibonacci number to compute, and after computing, the value is found in `t0`.
+* `naturalsum`
+* `multiplier`
+* `divider`
+
+If you have passed all of the above tests, your CPU should execute these applications with no issues!
+If you do not pass a test, you may need to dig into the debug output of the test.
+
+## Testing full applications
+
+You can run all of the applications at once with the following test.
+
+```
+sbt:dinocpu> testOnly dinocpu.SingleCycleApplicationsTesterLab2
+```
+
+To run a single application, you can use the following command:
+
+```
+sbt:dinocpu> testOnly dinocpu.SingleCycleApplicationsTesterLab2 -- -z <binary name>
+```
+
+# Feedback
+
+This time, instead of uploading a paper version to Gradescope, you will give feedback via a Google form.
+Note that the assignment will be out of 90 points and the last 10 points from the feedback will appear in Canvas sometime later.
+
+<link to google form here>
+
+# Grading
+
+Grading will be done automatically on Gradescope.
+See [the Submission section](#Submission) for more information on how to submit to Gradescope.
+
+| Name                  | Percentage                 |
+|-----------------------|----------------------------|
+| Each instruction type | 8% each (× 10 parts = 80%) |
```

Comment:
There's only 9 parts above, since the 10th part is the full programs. Are the branch control unit's tests considered part of the scoring here, and that's where the "10th" part comes from?